Feature/email support

tsteinruecken opened the pull request:
Implemented an Email-Queue and basic SendInBlue transport based on @sveneberth PR #24.

sveneberth reviewed (changes requested):
Hello @tsteinruecken,
thanks for you're extension. We're on a good way.

I've tested you're implementations with several cases and conditions and found a lot of bugs and vulnerabilities. I think you haven't tested this enough or your assumptions are not realistic (for example 100% API-Uptime).

In addition to problems marked in the code here are a few more:

- The filetypes of the attachments have to be checked in sendEmail. SIB does not support all filetypes (see https://apidocs.sendinblue.com/tutorial-sending-transactional-email/). Popular file-types like _.heic_, _.odt_ or _.svg_ cause a 400 Bad Request (Body: `{"code":"invalid_parameter","message":"Unsupported file format: heic"}`)
- I miss a hook-method or something like that, that allows me to store custom data according to the email. I use the concept of "eventHandlers" in customer projects. There can I pass a data-dict and a method-reference to the sendEMail-Method and after the SIB-Webhook calls an event for this email the method is executed with the data-dict. I think this full implementation must not be part of the server. But storing the content from a `queueEntityData`-argument in the sendEmail-method to the queueEntity and calling a method (referenced with `conf["viur.email.transportSuccessful"]`) with the key of the queueEntity after sending would be enough for me.
- The code, especially in the `sendEMail` method, is ugly and unreadable as hell. There is not even ONE blank line. Sorry, but I cannot understand how you can ignore basic concepts like [PEP 8](https://www.python.org/dev/peps/pep-0008/). We've talked about code refactoring and reformatting so many times in ViUR rounds... This should be indispensable for new code - especially as a maintainer.

tsteinruecken commented on email.py:
```diff
@@ -0,0 +1,265 @@
+				users.append(userSkel["name"])
+
+			if users:
+				ret = sendEMail(dests=users, template=os.linesep.join((subject, body)), *args, **kwargs)
```
Fixed

tsteinruecken commented on email.py:
```diff
@@ -0,0 +1,265 @@
+	assert queueEntity, "Email queue object went missing!"
+	if queueEntity["isSend"]:
+		return True
+	elif queueEntity["errorCount"] > 3:
```
Emails are being send through a custom queue (viur-emails). I've added documentation on how it should be configured (so that it retries only once an hour)

tsteinruecken commented on email.py:
```diff
@@ -0,0 +1,265 @@
+							 data=payload, headers=headers, method="POST")
+	try:
+		response = request.urlopen(reqObj)
+	except request.HTTPError as e:
```
HTTP-Error is correct. We can read() from that exception in this case to log the response from SIB. We don't care for anything else as we'll raise an exception anyway to have the queue retry.

tsteinruecken commented on email.py:
```diff
@@ -0,0 +1,265 @@
+	queueEntity["subject"] = subject
+	queueEntity["body"] = body
+	queueEntity["headers"] = headers
+	queueEntity["attachments"] = attachments
```
Fixed. That should be enough in most cases; however you're free to implement your own transport that generates files on the fly/uses GCS-Links etc.

tsteinruecken commented on email.py:
```diff
@@ -0,0 +1,265 @@
+		for file_name, file_content in attachments:
+			dataDict["attachment"].append({
+				"name": file_name,
+				"content": base64.b64encode(file_content)
```
Fixed.

tsteinruecken commented on email.py:
```diff
@@ -0,0 +1,265 @@
+def sendEMail(*,
+			  tpl: str = None,
+			  stringTemplate: str = None,
+			  skel: Union[None, Dict, "SkeletonInstance", List["SkeletonInstance"]] = None,
```
Nope. I will not import the skeleton system here as it pulls most of the ViUR System as it's dependency - which would prevent you from using the email service in there.

tsteinruecken commented on email.py:
```diff
@@ -0,0 +1,265 @@
+	:param headers: Specify headers for this email.
+	:param attachments: List of files ((filename, filecontent)-pairs) to be sent within the mail as attachments.
+
+	.. Warning:  As emails will be queued (and not send directly) you cannot exceed 1MB in total
```
That's absolutely fine. You want this limit to be documented upfront (instead of being surprised in production later on). The datastore will check the size later on (no need to do this twice).

tsteinruecken commented:
Regarding the other comments:
There's no need to check the filetypes. That's a purely SIB limitation (other transports may deliver heic or svgs just fine) - so the first time we can check these is inside the deferred call. So there's no advantage to letting SIB enforce this (they may also change their list of supported files in the future)

I've added the transportSuccessful hook as requested.

The code is fine. I've moved to normalizeToList function to the top-level, now it's fully compliant to our coding guideline :)
